com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/RestartFrameHandler.java

```diff
@@ -0,0 +1,110 @@
+import com.sun.jdi.StackFrame;
+import com.sun.jdi.ThreadReference;
+import com.sun.jdi.request.StepRequest;
+
+/**
+ * Support Eclipse's `Drop To Frame` action, which is restartFrame in VSCode's
+ * debug.
+ */
+public class RestartFrameHandler implements IDebugRequestHandler {
+
+    @Override
+    public List<Command> getTargetCommands() {
+        return Arrays.asList(Requests.Command.RESTARTFRAME);
+    }
+
+    @Override
+    public CompletableFuture<Response> handle(Command command, Arguments arguments, Response response, IDebugAdapterContext context) {
+        RestartFrameArguments restartFrameArgs = (RestartFrameArguments) arguments;
+        StackFrameReference stackFrameReference = (StackFrameReference) context.getRecyclableIdPool().getObjectById(restartFrameArgs.frameId);
+
+        if (stackFrameReference == null) {
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE,
+                    String.format("RestartFarme: cannot find the stack frame with frameID %s", restartFrameArgs.frameId));
+        }
+
+        if (canRestartFrame(context, stackFrameReference)) {
+            try {
+                ThreadReference reference = stackFrameReference.getThread();
+                popStackFrames(context, reference, stackFrameReference.getDepth());
+                stepin(context, reference);
+            } catch (DebugException de) {
+                context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, de.getMessage()));
+            }
+            return CompletableFuture.completedFuture(response);
+        } else {
+            context.getProtocolServer().sendEvent(new Events.UserNotificationEvent(NotificationType.ERROR, "Current stack frame doesn't support restart."));
+            context.getProtocolServer().sendEvent(new Events.StoppedEvent("restartframe", stackFrameReference.getThread().uniqueID()));
+            return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.RESTARTFRAME_FAILURE, "Failed to restart the selected stack frame.");
+        }
+    }
+
+    private boolean canRestartFrame(IDebugAdapterContext context, StackFrameReference frameReference) {
+        ThreadReference reference = frameReference.getThread();
+        StackFrame[] frames = context.getStackFrameManager().reloadStackFrames(reference);
+        // Current stack frame cannot be the top frame of the thread:
+        if (frames.length <= frameReference.getDepth() + 1) {
+            return false;
+        }
+
+        // Cannot restart frame involved with native call stacks:
+        for (int i = 0; i <= frameReference.getDepth() + 1; i++) {
+            if (StackFrameUtility.isNative(frames[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void popStackFrames(IDebugAdapterContext context, ThreadReference thread, int depth) throws DebugException {
+        while (depth >= 0) {
```

Comment:
It seems all stack frames up to the specified depth can be popped by the one invoke. See the doc https://docs.oracle.com/javase/7/docs/jdk/api/jpda/jdi/

Reply:
Updated.